pkg/network/http/driver_interface.go

```diff
@@ -0,0 +1,178 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build windows && npm
+// +build windows,npm
+
+package http
+
+/*
+#include <stdlib.h>
+#include <memory.h>
+*/
+import "C"
+import (
+	"runtime"
+	"sync"
+	"unsafe"
+
+	"github.com/DataDog/datadog-agent/pkg/network/driver"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"golang.org/x/sys/windows"
+)
+
+const (
+	httpReadBufferCount = 100
+)
+
+type FullHttpTransaction struct {
+	Txn             driver.HttpTransactionType
+	RequestFragment []byte
+}
+type httpDriverInterface struct {
+	driverHTTPHandle  driver.Handle
+	driverEventHandle windows.Handle
+
+	readMux     sync.Mutex
+	dataChannel chan []FullHttpTransaction
+	eventLoopWG sync.WaitGroup
+	closed      bool
+}
+
+func newDriverInterface(dh driver.Handle) (*httpDriverInterface, error) {
+	d := &httpDriverInterface{}
+	err := d.setupHTTPHandle(dh)
+	if err != nil {
+		return nil, err
+	}
+
+	d.dataChannel = make(chan []FullHttpTransaction)
+	return d, nil
+}
+
+func (di *httpDriverInterface) setupHTTPHandle(dh driver.Handle) error {
+
+	di.driverHTTPHandle = dh
+	// enable HTTP on this handle
+	settings := driver.HttpConfigurationSettings{
+		MaxTransactions:        driver.HttpBatchSize * 2,
+		NotificationThreshhold: driver.HttpBatchSize,
+		MaxRequestFragment:     driver.HttpBufferSize,
+	}
+
+	err := dh.DeviceIoControl(
+		driver.EnableHttpIOCTL,
+		(*byte)(unsafe.Pointer(&settings)),
+		uint32(driver.HttpSettingsTypeSize),
+		nil,
+		uint32(0), nil, nil)
+	if err != nil {
+		log.Warnf("Failed to enable http in driver %v", err)
+		return err
+	}
+	log.Infof("Enabled http in driver")
+
+	u16eventname, err := windows.UTF16PtrFromString("Global\\DDNPMHttpTxnReadyEvent")
+	di.driverEventHandle, err = windows.CreateEvent(nil, 1, 0, u16eventname)
+	if err != nil {
+		if err != windows.ERROR_ALREADY_EXISTS || di.driverEventHandle == windows.Handle(0) {
+			log.Warnf("Failed to create driver event handle %v", err)
+			return err
+		}
+		log.Infof("non-nil err, %v %v", di.driverEventHandle, err)
+	}
+	return nil
+}
+
+func (di *httpDriverInterface) readAllPendingTransactions() {
+	di.readMux.Lock()
+	defer di.readMux.Unlock()
+	count := int(0)
+	for {
+		txns, err := di.readPendingTransactions()
+		if err != nil {
+			log.Warnf("Error reading http transaction buffer: %v", err)
+			break
+		}
+		if txns == nil && err == nil {
+			// no transactions to read
+			break
+		}
+		count += len(txns)
+		di.dataChannel <- txns
```

Comment:
Should we remove this channel and have this function read all the transactions available, return them as a batch & then we can process them synchronously? This should help mitigate the race condition which you mentioned